# **Ejercicios de pair programming Módulo 3 Sprint 1**
## **Regreción Lineal: Lección 7 - ANOVA**

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.


Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.


📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [1]:
# Tratamiento de los datos
# ========================
import numpy as np
import pandas as pd

# Librerías para la visualización de los datos
# ============================================
import matplotlib.pyplot as plt 
import seaborn as sns

# Estadísticos
# ============
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

In [2]:
df = pd.read_csv("datos/soci_econ_country_profiles_estandarizado.csv", index_col = 0)
df.sample(5)

region  restaurant_price_index  population_2017  \
country                                                                      
Slovenia           SouthernEurope                   43.13        -0.285851   
Slovakia            EasternEurope                   33.17        -0.228500   
Greece             SouthernEurope                   54.62        -0.131235   
Republic of Korea     EasternAsia                   44.62         0.546857   
Turkey                WesternAsia                   26.45         1.053664   

                   employment_agriculture  employment_industry  \
country                                                          
Slovenia                         0.240166             0.798611   
Slovakia                        -0.223602             1.479167   
Greece                           0.571429            -1.229167   
Republic of Korea               -0.082816             0.118056   
Turkey                           1.126294             0.506944   

                   employment_services  education  quality_of_life_index  \
country                                                                    
Slovenia                     -0.350072   0.352941               0.533877   
Slovakia                     -0.309900  -0.258824               0.178425   
Greece                        0.258250  -2.031603               0.112767   
Republic of Korea             0.154950   0.164706               0.332712   
Turkey                       -0.843615   0.023529              -0.177338   

                   purchasing_power_index  safety_index  cost_of_living  \
country                                                                   
Slovenia                         0.168545      0.877752        0.099747   
Slovakia                        -0.111428      0.538280       -0.141730   
Greece                          -0.246638     -0.023259        0.175896   
Republic of Korea                0.670128      0.810190        0.741658   
Turkey                          -0.300639     -0.065901       -0.324140   

                   consumer_price_index  human_capital_index  inflation  \
country                                                                   
Slovenia                      -0.308375             0.477745  -0.268650   
Slovakia                      -0.199931            -0.080119  -0.322319   
Greece                        -0.622682            -0.157270  -0.409677   
Republic of Korea             -0.028438             0.816024  -0.032585   
Turkey                         3.040076            -0.483680   4.181595   

                   urban_population  population_age_distribution_0_14  \
country                                                                 
Slovenia                  -0.754029                         -0.227405   
Slovakia                  -0.772966                         -0.180758   
Greece                     0.132994                         -0.320700   
Republic of Korea          0.233809                         -0.402332   
Turkey                    -0.015019                          0.938776   

                   population_age_distribution_60_plus  \
country                                                  
Slovenia                                      0.384906   
Slovakia                                      0.045283   
Greece                                        0.400000   
Republic of Korea                            -0.083019   
Turkey                                       -0.694340   

                   international_migrant_stock  restaurant_price_index_normal  
country                                                                        
Slovenia                              0.213018                       4.065429  
Slovakia                             -0.426036                       3.761357  
Greece                                0.205128                       4.341706  
Republic of Korea                    -0.481262                       4.104996  
Turkey                           

In [3]:
# iniciamos la ANOVA para la variable respuesta normalizada 'restaurant_price_index_normal'
lm = ols('''restaurant_price_index_normal ~ region + population_2017 + employment_agriculture +
       employment_industry + employment_services + education +
       quality_of_life_index + purchasing_power_index + safety_index +
       cost_of_living + consumer_price_index + human_capital_index +
       inflation + urban_population + population_age_distribution_0_14 +
       population_age_distribution_60_plus + international_migrant_stock''', 
         data=df).fit()
# resultados obtenidos en el ANOVA
sm.stats.anova_lm(lm)

df     sum_sq   mean_sq           F  \
region                               13.0  15.935057  1.225774   46.662113   
population_2017                       1.0   0.437499  0.437499   16.654467   
employment_agriculture                1.0   2.649718  2.649718  100.868095   
employment_industry                   1.0   0.235514  0.235514    8.965425   
employment_services                   1.0   0.253964  0.253964    9.667764   
education                             1.0   0.004721  0.004721    0.179726   
quality_of_life_index                 1.0   0.009399  0.009399    0.357782   
purchasing_power_index                1.0   0.057461  0.057461    2.187399   
safety_index                          1.0   0.020066  0.020066    0.763877   
cost_of_living                        1.0   2.797179  2.797179  106.481549   
consumer_price_index                  1.0   0.010235  0.010235    0.389632   
human_capital_index                   1.0   0.043117  0.043117    1.641358   
inflation                             1.0   0.156283  0.156283    5.949305   
urban_population                      1.0   0.171459  0.171459    6.527005   
population_age_distribution_0_14      1.0   0.000239  0.000239    0.009087   
population_age_distribution_60_plus   1.0   0.071716  0.071716    2.730056   
international_migrant_stock           1.0   0.012273  0.012273    0.467195   
Residual                             36.0   0.945689  0.026269         NaN   

                                           PR(>F)  
region                               1.496188e-18  
population_2017                      2.378539e-04  
employment_agriculture               5.534266e-12  
employment_industry                  4.951272e-03  
employment_services                  3.655035e-03  
education                            6.741312e-01  
quality_of_life_index                5.534854e-01  
purchasing_power_index               1.478418e-01  
safety_index                         3.879111e-01  
cost_of_living                       2.666975e-12  
consumer_price_index                 5.364290e-01  
human_capital_index                  2.083308e-01  
inflation                            1.978250e-02  
urban_population                     1.499748e-02  
population_age_distribution_0_14     9.245854e-01  
population_age_distribution_60_plus  1.071728e-01  
international_migrant_stock          4.986566e-01  
Residual                                      NaN

In [4]:
# sacamos el resumen de la recta que ajusta nuestro modelo
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                  OLS Regression Results                                 
=========================================================================================
Dep. Variable:     restaurant_price_index_normal   R-squared:                       0.960
Model:                                       OLS   Adj. R-squared:                  0.928
Method:                            Least Squares   F-statistic:                     30.02
Date:                           Tue, 16 May 2023   Prob (F-statistic):           1.13e-17
Time:                                   14:47:33   Log-Likelihood:                 46.451
No. Observations:                             66   AIC:                            -32.90
Df Residuals:                                 36   BIC:                             32.79
Df Model:                                     29                                         
Covariance Type:                       nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               3.1332      1.319      2.375      0.023       0.458       5.809
region[T.EasternAsia]                  -0.3768      0.253     -1.489      0.145      -0.890       0.136
region[T.EasternEurope]                 0.0808      0.222      0.364      0.718      -0.370       0.531
region[T.NorthernAfrica]                0.8828      0.395      2.234      0.032       0.081       1.684
region[T.NorthernAmerica]               0.1873      0.251      0.747      0.460      -0.321       0.696
region[T.NorthernEurope]                0.3170      0.223      1.423      0.163      -0.135       0.769
region[T.Oceania]                       0.0779      0.244      0.320      0.751      -0.416       0.572
region[T.South-easternAsia]            -0.4003      0.234     -1.714      0.095      -0.874       0.073
region[T.SouthAmerica]                 -0.0193      0.199     -0.097      0.923      -0.423       0.384
region[T.SouthernAfrica]                0.7143      0.313      2.284      0.028       0.080       1.349
region[T.SouthernAsia]                  0.1426      0.254      0.561      0.579      -0.373       0.658
region[T.SouthernEurope]                0.0814      0.234      0.348      0.730      -0.393       0.556
region[T.WesternAsia]                   0.1749      0.201      0.869      0.391      -0.234       0.583
region[T.WesternEurope]                 0.2436      0.239      1.021      0.314      -0.240       0.727
population_2017                         0.0110      0.008      1.398      0.171      -0.005       0.027
employment_agriculture                  3.8875      6.418      0.606      0.549      -9.129      16.904
employment_industry                     2.3022      3.818      0.603      0.550      -5.440      10.045
employment_services                     5.6169      9.250      0.607      0.548     -13.144      24.377
education                               0.0012      0.002      0.492      0.626      -0.004       0.006
quality_of_life_index                   0.0773      0.108      0.719      0.477      -0.141       0.295
purchasing_power_index                 -0.2593      0.098     -2.645      0.012      -0.458      -0.060
safety_index                           -0.0626      0.061     -1.031      0.310      -0.186       0.061
cost_of_living                          0.6338      0.078      8.082      0.000       0.475       0.793
consumer_price_index                    0.0007      0.027      0.025      0.980      -0.054       0.055
human_capital_index                     0.2686      0.105      2.547      0.015       0.055       0.482
inflation                              -0.0758    

CONCLUSIONES:

- El R-Square de nuestros datos es muy alto, llegando a 0.96, es decir, los datos sugieren que nuestro modelo de predición es muy fiable porque explica el 96% de la variación de la variable respuesta. Sin embargo, estas conclusiones son fictícias ya que nuestros datos no siguen una distribución normal. 
  
- Las variables que afectan la variable respuesta (pvalue <= 0.05) en nuetros datos son: 'region', 'purchasing_power_index', 'cost_of_living', 'human_capital_index', 'inflation' y 'urban_population'. Estas son las que deberemos usar para el modelo de regresión lineal. 
  
- De las columnas que afectan nuestra VR sacamos los coef: 'region' (con distintos valores), 'purchasing_power_index' (coef: -0.2593), 'cost_of_living' (coef 0.6338), 'human_capital_index' (coef 0.2686), 'inflation' (coef -0.0758) y 'urban_population' (coef 0.2021). En el caso de la variable 'purchasing_power_index' (coef: -0.2593), por cada unidad que aumenta 'purchasing_power_index', la variable respuesta disminuye 0.2593. 